plugin/pkg/admission/eventratelimit/admission.go

```diff
@@ -0,0 +1,219 @@
+*/
+
+package eventratelimit
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/golang/groupcache/lru"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/kubernetes/pkg/api"
+	eventratelimitapi "k8s.io/kubernetes/plugin/pkg/admission/eventratelimit/apis/eventratelimit"
+	"k8s.io/kubernetes/plugin/pkg/admission/eventratelimit/apis/eventratelimit/validation"
+)
+
+const (
+	defaultCacheSize = 4096
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register("EventRateLimit",
+		func(config io.Reader) (admission.Interface, error) {
+			// load the configuration provided (if any)
+			configuration, err := LoadConfiguration(config)
+			if err != nil {
+				return nil, err
+			}
+			// validate the configuration (if any)
+			if configuration != nil {
+				if errs := validation.ValidateConfiguration(configuration); len(errs) != 0 {
+					return nil, errs.ToAggregate()
+				}
+			}
+			return newEventRateLimit(configuration, realClock{})
+		})
+}
+
+// eventRateLimitAdmission implements an admission controller that can enforce event rate limits
+type eventRateLimitAdmission struct {
+	*admission.Handler
+	// limitEnforcers is the collection of limit enforcers. There is one limit enforcer for each
+	// active limit type. As there are 4 limit types, the length of the array will be at most 4.
+	// The array is read-only after construction.
+	limitEnforcers []*limitEnforcer
+}
+
+// realClock implements flowcontrol.Clock in terms of standard time functions.
+type realClock struct{}
+
+// Now is identical to time.Now.
+func (realClock) Now() time.Time {
+	return time.Now()
+}
+
+// Sleep is identical to time.Sleep.
+func (realClock) Sleep(d time.Duration) {
+	time.Sleep(d)
+}
+
+// limitEnforcer enforces a single type of event rate limit, such as server, namespace, or source+object
+type limitEnforcer struct {
+	// type of this limit
+	limitType eventratelimitapi.LimitType
+	// cache for holding the rate limiters
+	cache cache
+	// function for selecting the cache key with which an event is associated
+	keySelector func(admission.Attributes) interface{}
+}
+
+func newLimitEnforcer(config eventratelimitapi.Limit, clock flowcontrol.Clock) (*limitEnforcer, error) {
+	rateLimiterFactory := func() flowcontrol.RateLimiter {
+		return flowcontrol.NewTokenBucketRateLimiterWithClock(config.QPS, int(config.Burst), clock)
+	}
+
+	limitEnforcer := &limitEnforcer{
+		limitType: config.Type,
+	}
+
+	if config.Type == eventratelimitapi.ServerLimitType {
+		limitEnforcer.cache = &singleCache{
+			rateLimiter: rateLimiterFactory(),
+		}
+	} else {
+		cacheSize := int(config.CacheSize)
+		if cacheSize == 0 {
+			cacheSize = defaultCacheSize
+		}
+		limitEnforcer.cache = &lruCache{
+			rateLimiterFactory: rateLimiterFactory,
+			cache:              lru.New(cacheSize),
+		}
+	}
+
+	switch t := config.Type; t {
```

Comment:
nit: once restructuring the above, do a `var keyfunc func(admission.Attributes) interface{}`, and create/return all at once instead of mutating your value going through.